lib/cloud/clients.go

```diff
@@ -942,3 +1058,90 @@ func (c *TestCloudClients) GetAzureRunCommandClient(subscription string) (azure.
 func (c *TestCloudClients) Close() error {
 	return nil
 }
+
+// AWSSessionCacheKeyBuilder is a helper struct wrapper around a string builder
+// that is used to efficiently build cache keys for a given region and, optionally,
+// a chain of AWS assumed-role sessions.
+type AWSSessionCacheKeyBuilder struct {
+	sb        *strings.Builder
+	roleCount int
+}
+
+// NewAWSSessionCacheKeyBuilder constructs a new cache key builder intialized
+// with a specific AWS region.
+func NewAWSSessionCacheKeyBuilder(region string) AWSSessionCacheKeyBuilder {
+	sb := strings.Builder{}
+	_, _ = sb.WriteString(region) // infalliable.
+	return AWSSessionCacheKeyBuilder{sb: &sb}
+}
+
+// AddRole adds an AWS IAM role to the cache key builder.
+func (c *AWSSessionCacheKeyBuilder) AddRole(awsRole services.AssumeRole) {
+	// Role ARN can potentially come from user input via --db-user for some
+	// protocols.
+	// Out of paranoia for crafted user input, I build the key in a way that
+	// will avoid fetching an AWS session from cache that the user should not
+	// have access to.
+	key := fmt.Sprintf(":Role[%d]:ARN[%v]:ExternalID[%v]", c.roleCount, awsRole.RoleARN, awsRole.ExternalID)
```

Comment:
i'm not sure if the `arn.Parse` will accept that since it has extra semicolons in it, but assuming it does:

`<region>:Role[0]:ARN[<valid-arn>:Role[1]:ARN[<another-arn>]:ExternalID[<someid>]]:ExternalID[<id-from-config>]`

Assume that the configuration is:
```
assume_role_arn: arn:aws:iam::123456789012:role/role-1
external_id: someUniqueID
```

Best I can think of is
```
--db-user=arn:aws:iam::123456789012:role/role-2]:Role[2]:ARN[arn:aws:iam::123456789012:role/role-3
```

cache key: 
```
us-west-1:Role[0]:ARN[arn:aws:iam::123456789012:role/role-1]:ExternalID[someUniqueID]:Role[1]:ARN[arn:aws:iam::123456789012:role/role-2]:Role[2]:ARN[arn:aws:iam::123456789012:role/role-3]:ExternalID[]
```

which still isn't right since it's missing an `ExternalID[]` in the key

Reply:
refactored to remove use of cache key builder, although I haven't removed the builder itself yet. I don't think this is an issue regardless, so I'm marking resolved